### ISSUE TITLE
filepath issue

### ISSUE THREAD

**Issue by jas99** [yarnpkg/yarn#2599]
Issue with running something like
`yarn add /d/repo/github.com/jas99/node-bunyan.git`

as path contains folder with contain special character '.'

**Error**
```
[1/5] Resolving packages...
error Command failed.
Exit code: 128
Command: git
Arguments: clone file:/d/repo/github.com/jas99/node-bunyan.git /home/jas/.cache/yarn/.tmp/095537e455e8896b1a501aa14c31204f
Directory: /d/repo/gitlab.com/fit3d/viz3d.git
Output:
Cloning into '/home/jas/.cache/yarn/.tmp/095537e455e8896b1a501aa14c31204f'...
ssh: Could not resolve hostname file: Name or service not known
fatal: Could not read from remote repository.

```

**Comment by markstos** [comment]
@jas99 You are mixing file path syntax with git repo syntax. Have you found documentation somewhere which states that `yarn` or `npm` supports installing from `git` repos using the `file:/` prefix?

**Comment by jas99** [comment]
@markstos thx for reply
Sorry I meant
`yarn add file:/d/repo/github.com/jas99/node-bunyan.git`

**Comment by markstos** [comment]
@jas99 You are **_still_** mixing file path syntax with git repo syntax. Have you found documentation somewhere which states that yarn or npm supports installing from git repos using the file:/ prefix?

**Comment by jas99** [comment]
Hey @markstos 

https://yarnpkg.com/lang/en/docs/cli/add/
> yarn add file:/path/to/local/folder installs a package that is on your local file system. This is useful to test out other packages of yours that haven’t been published to the registry.

and btw it works fine if 'dot' is not thr in path of folder

Thx

**Comment by markstos** [comment]
@jas99, yes, "file:/" paths work and remote Git URLs work, but you have mixed the two, and I find no evidence that's supported. Let's look the output you provided:

> Command: git
> Arguments: clone file:/d/repo/github.com/jas99/node-bunyan.git 
> ...
> ssh: Could not resolve hostname file: Name or service not known

So it is clearly trying to call the command `git` with the arguments of `clone` and a `file:/` path.  Then it trying to resolve `file:` as a hostname to connect to over SSH.

Because of the ".git" extension it appears that the remote Git URL is being logic is being used. If you don't want that, try not storing your package directory in a folder that ends in `.git`.

**Comment by jas99** [comment]
@markstos Not able to store folder as .git folder on disk was the reason I created this issue. Otherwise everything works. I think some people like me follow this convention and since file protocol was explicitly mentioned in yarn add; I see no reason why using .git folder using file protocol should not work.

Thx

**Comment by markstos** [comment]
@jas99, I read the source code to understand how it decides whether to treat a repository as a "file" path vs a "git" path. The code has a list of resolvers which are checked in an apparently arbitrary order to see which one resolver matches the "repository" pattern that you passed. 

As it happens, the "git" resolver is checked first, and the git resolver checks for a ".git" suffix, so a repository that starts with "file:/" and ends with ".git" gets interpreted as a ".git" repo.

I think it would be reasonable for the logic to be reserved so the "file" resolver is checked first, so a URL of "file:/path/to/foo.git" would be interpreted as a file path, not a remote git repo.  I think the "protocol" prefix is a stronger signal of the correct resolver to use than the ".git" suffix. As @jas99 points out, some people could name directories with the ".git" extension that available locally.

If core devs agree, the fix would be to check the resolvers in an explicit order, then make sure the "file:" resolver pattern is checked before the "git" resolver pattern is tried.
